nodejs/aws-infra/experimental/cluster.ts

```diff
@@ -0,0 +1,141 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as aws from "@pulumi/aws";
+import * as pulumi from "@pulumi/pulumi";
+
+import { Network } from "./../network";
+
+import * as utils from "../utils";
+
+/**
+ * A Cluster is a general purpose ECS cluster configured to run in a provided Network.
+ */
+export class Cluster extends pulumi.ComponentResource {
+    public readonly instance: aws.ecs.Cluster;
+
+    /**
+     * The network in which to create this cluster.
+     */
+    public readonly network: Network;
+    /**
+     * Security groups associated with this this ECS Cluster.
+     */
+    public readonly instanceSecurityGroups: aws.ec2.SecurityGroup[];
+
+    constructor(name: string, args: ClusterArgs, opts: pulumi.ComponentResourceOptions = {}) {
+        super("aws-infra:x:Cluster", name, {
+            ...args,
+        }, opts);
+
+        // First create an ECS cluster.
+        const parentOpts = { parent: this };
+        const instance = args.instance || new aws.ecs.Cluster(name, args, parentOpts);
+
+        const network = args.network || Network.getDefault();
+
+        // IDEA: Can we re-use the network's default security group instead of creating a specific
+        // new security group in the Cluster layer?  This may allow us to share a single Security Group
+        // across both instance and Lambda compute.
+        const instanceSecurityGroups = args.instanceSecurityGroups || [new aws.ec2.SecurityGroup(name, {
+            vpcId: network.vpcId,
+            ingress: Cluster.defaultSecurityGroupIngress(),
+            egress: Cluster.defaultSecurityGroupEgress(),
+            tags: { Name: name },
+        }, parentOpts)];
+
+        this.instance = instance;
+        this.network = network;
+        this.instanceSecurityGroups = instanceSecurityGroups;
+
+        this.registerOutputs({
+            instance,
+            network,
+            instanceSecurityGroups,
+        });
+    }
+
+    public static defaultSecurityGroupEgress() {
+        return [{
+            fromPort: 0,
+            toPort: 0,
+            protocol: "-1",  // all
+            cidrBlocks: [ "0.0.0.0/0" ],
+        }];
+    }
+
+    public static defaultSecurityGroupIngress() {
+        return [
+            // Expose SSH
+            {
+                fromPort: 22,
+                toPort: 22,
+                protocol: "TCP",
+                cidrBlocks: [ "0.0.0.0/0" ],
+            },
+            // Expose ephemeral container ports to Internet.
+            // TODO: Limit to load balancer(s).
```

Comment:
Should we file a work item so we don't forget to do this? If it's not in the initial version, fixing this later would be a breaking change. It's also something we probably want to consider exposing arguments to control.

Reply:
we're discussing right now the best patterns we canhave for sensible defaults.  Right now this PR demosntrates an approach whereby:

1. we (pulumi) can have a default in case users do not want to supply values.
2. users can opt out entirely from our default and supply their own.

What is missing is the best way to allow users to *interact* with our defaults.  i.e. how does a user say: "i like your default, but i want to tweak this one part".  What this approach provide is a way for them to at least get at the data so they don't need to necessarily repeat it on their own side.

1. We need to decide if waht we have here is sufficient.
2. We need to make sure what what we have here does'nt block future work to make things more flexible if we decide it isn't.  

This is on the list for @lukehoban and i to discuss as part of the overall design/patterns here for this layer.

Reply:
Decision: open issues for all these decisions.  We will want to ensure on a resource-by-resource basis that these values are what we want.